pytorch_lightning/callbacks/model_summary.py

```diff
@@ -46,11 +46,11 @@ class ModelSummary(Callback):
         >>> trainer = Trainer(callbacks=[ModelSummary(max_depth=1)])
     """
 
-    def __init__(self, max_depth: Optional[int] = 1):
+    def __init__(self, max_depth: int = 1) -> None:
         self._max_depth: int = max_depth
 
     def on_pretrain_routine_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
-        if trainer.is_global_zero and self._max_depth is not None and not trainer.testing:
+        if trainer.is_global_zero and self._max_depth and not trainer.testing:
```

Comment:
this can be a separate PR, but creating the summary should not be gated by `is_global_zero` here. This will break for DeepSpeed/FSDP or any model with parameters on different devices in the forward pass if the LightningModule sets an `example_input_array`. `forward` needs to act like a collective, so it must run on all ranks at the same time. it's one of the points listed in https://github.com/PyTorchLightning/pytorch-lightning/issues/9043

It is safer to generate the summary data unconditionally on all ranks and then let the logging of the summarization be determined by the particular rank id

Reply:
yes I agree. only guard the logging instruction

Reply:
also the condition  not trainer.testing here is always true, the hook here only runs for training :)